### DISCUSSION TITLE
Need help debugging unreachable code

### DISCUSSION THREAD

**Discussion by gmlewis** [titzer/virgil#256] [Q&A] [answered]
While trying to learn Virgil, I've generated a .wasm with unreachable code and am interested in getting tips on how to debug this.
Here's the repo, branch, and directory: https://github.com/gmlewis/virgil/tree/extism-pdk/apps/Extism

To recreate, you first need the [Extism CLI](https://github.com/extism/cli):

```bash
$ ./build.sh
...
$ ./run.sh
+ extism call greet/Greet.wasm greet --wasi --input Benjamin
Hello, Benjamin!

+ extism call count-vowels/CountVowels.wasm count_vowels --wasi --input 'Once upon a dream'
{"count":7,"total":7,"vowels":"aeiouAEIOU"}

+ extism call http-get/HttpGet.wasm http_get --wasi '--allow-host=*.typicode.com'
Error: wasm error: unreachable
wasm stack trace:
	.$40(i32,i32,i32,i32) i32
	.$31(i32,i32,i32) i32
	.$20(i32) i32
	.$14(i32,i32) i32
	.$10() i32
```

I've used `wasm2wat` but I'm thinking there must be a better way, and would love to get advice.
Thank you!

**Comment by titzer** [comment, marked as answer]
One thing is to add `-symbols` when compiling; that will give you source-level function names. The Wasm target doesn't have source lines yet, a stacktrace with just function names is the best you can do for now. Any `unreachable` that you hit is going to be a safety check or an out-of-memory, so look for places where a `NullCheckException` or a `BoundsCheckException`, etc could be triggered.

**Reply by gmlewis** [reply]
Thank you, @titzer !

**Comment by gmlewis** [comment]
That was the key, @titzer - thank you!

Next question - now that I've got all 3 example Extism plugins working, are you interested in a PR into your main repo, or would you like me to keep my Extism Virgil PDK separate in my own fork?

**Reply by titzer** [reply]
Sure, happy to have an example Extism plugin, if it doesn't pull in any dependencies. It could live under `apps/`. Does it make sense to collapse them all into one directory or prefer to keep separate subdirectories?

**Reply by gmlewis** [reply]
OK, sounds good. Yes, I think one directory is fine. 
I'll send you a PR then. Thanks.

**Reply by gmlewis** [reply]
I've made one here: https://github.com/titzer/virgil/pull/258
